Technical author review of getting_started.md (#242)
ARMmbed/mbed-crypto: docs/getting_started.md

### Signing a message using RSA

Mbed Crypto supports encrypting, decrypting, signing and verifying messages using public key signature algorithms, such as RSA or ECDSA.

> **Prerequisites to performing asymmetric signature operations:**

Reviewer: I see a mix of `Prerequisites to` and `Prerequisites for` in this guide. Should we pick one usage and use it consistently?

* Initialize the library with a successful call to `psa_crypto_init()`.
* Have a valid key with appropriate attributes set:
    * Usage flag `PSA_KEY_USAGE_SIGN` to allow signing.
    * Usage flag `PSA_KEY_USAGE_VERIFY` to allow signature verification.
    * Algorithm set to the desired signature algorithm.

> This example shows how to sign a given hash using RSA, call `psa_asymmetric_sign()` and get the output buffer that contains the signature:

Reviewer: How do you know that `uint8_t hash[32]` is a message and not a pre-computed hash?

> This example shows how to sign a given hash using RSA, call `psa_asymmetric_sign()` and get the output buffer that contains the signature:

Reviewer 2: It wasn't `uint8_t hash[32]` before, it was `uint8_t hash[] = "INPUT_FOR_SIGN"`. That's obviously not a pre-computed hash.

> This example shows how to sign a given hash using RSA, call `psa_asymmetric_sign()` and get the output buffer that contains the signature:

Reviewer: Does the function `psa_asymmetric_sign` assume the input is already hashed or does it do the hashing and signing for you all at once? The example is obviously wrong if it's the latter, and the name `hash` for the variable is misleading.

> This example shows how to sign a given hash using RSA, call `psa_asymmetric_sign()` and get the output buffer that contains the signature:

Reviewer 2: `psa_asymmetric_sign` takes a hash as input.

> This example shows how to sign a given hash using RSA, call `psa_asymmetric_sign()` and get the output buffer that contains the signature:

Reviewer: The doxygen says "hash or message", which is confusing. Should it be already hashed or not... Maybe this depends on the algorithm passed in?

    * Usage flag `PSA_KEY_USAGE_VERIFY` to allow signature verification.
    * Algorithm set to the desired signature algorithm.

This example shows how to sign a hash that has already been calculated:
```C
    psa_status_t status;
    psa_key_attributes_t attributes = PSA_KEY_ATTRIBUTES_INIT;
    uint8_t key[] = RSA_KEY;
    uint8_t hash[32] = {0x50, 0xd8, 0x58, 0xe0, 0x98, 0x5e, 0xcc, 0x7f,

>                     0x60, 0x41, 0x8a, 0xaf, 0x0c, 0xc5, 0xab, 0x58,

Author: Thanks, @gilles-peskine-arm!
Updated.

### Using symmetric ciphers

Mbed Crypto supports encrypting and decrypting messages using various symmetric cipher algorithms (both block and stream ciphers).

**Prerequisites to working with the symmetric cipher API:**
* Initialize the library with a successful call to `psa_crypto_init()`.

> * Configure the key policy accordingly (set `PSA_KEY_USAGE_ENCRYPT` to allow encryption or `PSA_KEY_USAGE_DECRYPT` to allow decryption).

Reviewer 2: This was written for an earlier version of the API which made the word “policy” more prominent (it used to involve a call to `psa_set_key_policy`, but this function no longer exists). Now the important word is “usage” or “usage flags” (the function call is `psa_set_key_usage_flags`). The next line also uses obsolete terminology: we no longer have “key slots”. Suggested wording:

> * Have a handle to a symmetric key. This key's usage flags must include `PSA_KEY_USAGE_ENCRYPT` to allow encryption or `PSA_KEY_USAGE_DECRYPT` to allow decryption.

* Have a valid key in the key slot.

**To encrypt a message with a symmetric cipher:**
1. Allocate an operation (`psa_cipher_operation_t`) structure to pass to the cipher functions.

> 1. Call `psa_cipher_encrypt_setup()` to initialize the operation structure and specify the algorithm and the key to be used.

Reviewer 2: Preexisting: there's a missing step of initialization between “allocate” and “setup”. `psa_cipher_encrypt_setup` used to initialize and set up, but it now only sets up.

```suggestion
1. Initialize the operation structure to zero or to `PSA_CIPHER_OPERATION_INIT`.
1. Call `psa_cipher_encrypt_setup()` to specify the algorithm and the key to be used.
```

The same applies to the decryption process below.

1. Call either `psa_cipher_generate_iv()` or `psa_cipher_set_iv()` to generate or set the initialization vector (IV). We recommended calling `psa_cipher_generate_iv()`, unless you require a specific IV value.

> 1. Call `psa_cipher_update()` one or more times, passing the whole message or a fragment of the message on each call.

Reviewer 2: Preexisting: this looks like it's saying you can call `psa_cipher_update` multiple times and pass it the whole message each time. Suggestion:

> Call `psa_cipher_update()` one or more times, passing successive fragments of the message.

Alternatively:

> Call `psa_cipher_update()` with the message to encrypt. You may call this function multiple times, passing successive fragments of the message on successive calls.

The same applies to the decryption process below, and also to the similar processes for hashes and MAC.

Reviewer: @GuyWi This still needs revising

1. Call `psa_cipher_finish()` to end the operation and output the encrypted message.

> This example shows how to encrypt data using an Advanced Encryption Standard (AES) key in cipher block chain (CBC) mode with no padding (assuming all prerequisites have been fulfilled):

Reviewer: Replace "cipher block chain" with "Cipher Block Chaining".

> This example shows how to encrypt data using an Advanced Encryption Standard (AES) key in Cipher Block Chaining (CBC) mode with no padding (assuming all prerequisites have been fulfilled):

Reviewer 2: “AES” and “CBC” are better known than their etymological expansion. This should be “AES (Advanced Encryption Standard)” and “CBC (Cipher Block Chaining)”, if you bother to expand the acronym at all.

### Using symmetric ciphers

Mbed Crypto supports encrypting and decrypting messages using various symmetric cipher algorithms (both block and stream ciphers).

**Prerequisites to working with the symmetric cipher API:**
* Initialize the library with a successful call to `psa_crypto_init()`.
* Have a handle to a symmetric key. This key's usage flags must include `PSA_KEY_USAGE_ENCRYPT` to allow encryption or `PSA_KEY_USAGE_DECRYPT` to allow decryption.

**To encrypt a message with a symmetric cipher:**
1. Allocate an operation (`psa_cipher_operation_t`) structure to pass to the cipher functions.
1. Initialize the operation structure to zero or to `PSA_CIPHER_OPERATION_INIT`.
1. Call `psa_cipher_encrypt_setup()` to specify the algorithm and the key to be used.
1. Call either `psa_cipher_generate_iv()` or `psa_cipher_set_iv()` to generate or set the initialization vector (IV). We recommend calling `psa_cipher_generate_iv()`, unless you require a specific IV value.
1. Call `psa_cipher_update()` one or more times, passing the whole message or a fragment of the message on each call.
1. Call `psa_cipher_finish()` to end the operation and output the encrypted message.

> This example shows how to encrypt data using an AES (Advanced Encryption Standard) key in CBC (Cipher Block Chaining)) mode with no padding (assuming all prerequisites have been fulfilled):

Reviewer: Looks like there is an extra `)` here. Please remove.

    mbedtls_psa_crypto_free();
```

**To decrypt a message with a symmetric cipher:**
1. Allocate an operation (`psa_cipher_operation_t`) structure to pass to the cipher functions.
1. Call `psa_cipher_decrypt_setup()` to initialize the operation structure and to specify the algorithm and the key to be used.
1. Call `psa_cipher_set_iv()` with the IV for the decryption.

> 1. Call `psa_cipher_update()` one or more times, passing the whole message or only a fragment of the message each time.

Reviewer: Suggestion:
"passing the whole message or a fragment of the message on each call."

    mbedtls_psa_crypto_free();
```

**To decrypt a message with a symmetric cipher:**
1. Allocate an operation (`psa_cipher_operation_t`) structure to pass to the cipher functions.
1. Call `psa_cipher_decrypt_setup()` to initialize the operation structure and to specify the algorithm and the key to be used.
1. Call `psa_cipher_set_iv()` with the IV for the decryption.

> 1. Call `psa_cipher_update()` one or more times, passing the whole message or a fragment of the message on each call.

Reviewer: This needs revising in the same way as encrypt above.

* A call to `psa_cipher_finish()` succeeds or fails.

After an implicit or explicit call to `psa_cipher_abort()`, the operation structure is invalidated; in other words, you cannot reuse the operation structure for the same operation. You can, however, reuse the operation structure for a different operation by calling either `psa_cipher_encrypt_setup()` or `psa_cipher_decrypt_setup()` again.

You must call `psa_cipher_abort()` at some point for any operation that is initialized successfully (by a successful call to `psa_cipher_encrypt_setup()` or `psa_cipher_decrypt_setup()`).

Making multiple sequential calls to `psa_cipher_abort()` on an operation that is terminated (either implicitly or explicitly) is safe and has no effect.

### Hashing a message

Mbed Crypto lets you compute and verify hashes using various hashing
algorithms.

**Prerequisites to working with the hash APIs:**
* Initialize the library with a successful call to `psa_crypto_init()`.

**To calculate a hash:**
1. Allocate an operation structure (`psa_hash_operation_t`) to pass to the hash functions.

> 1. Call `psa_hash_setup()` to initialize the operation structure and specify the hash algorithm.

Reviewer 2: Preexisting: there's a missing step of initialization between “allocate” and “setup”. `psa_cipher_encrypt_setup` used to initialize and set up, but it now only sets up.

```suggestion
1. Initialize the operation structure to zero or to `PSA_HASH_OPERATION_INIT`.
1. Call `psa_hash_setup()` to specify the hash algorithm.
```

1. Call `psa_hash_update()` one or more times, passing the whole message or a fragment of the message on each call.
1. Call `psa_hash_finish()` to calculate the hash, or `psa_hash_verify()` to compare the computed hash with an expected hash value.

> This example shows how to calculate the `SHA-256` hash of a message:

Reviewer 2: SHA-256 should not be formatted as code.

Reviewer: @GuyWi This needs fixing still. Remove the backticks around SHA-256.

Mbed Crypto implicitly calls `psa_hash_abort()` when:
1. A call to `psa_hash_update()` fails (returning any status other than `PSA_SUCCESS`).
1. A call to `psa_hash_finish()` succeeds or fails.
1. A call to `psa_hash_verify()` succeeds or fails.

After an implicit or explicit call to `psa_hash_abort()`, the operation structure is invalidated; in other words, you cannot reuse the operation structure for the same operation. You can, however, reuse the operation structure for a different operation by calling `psa_hash_setup()` again.

You must call `psa_hash_abort()` at some point for any operation that is initialized successfully (by a successful call to `psa_hash_setup()`) .

Making multiple sequential calls to `psa_hash_abort()` on an operation that has already been terminated (either implicitly or explicitly) is safe and has no effect.

### Generating a random value

Mbed Crypto can generate random data.

> **Prerequisites to random generation:**

Reviewer: Suggestion: Use "random data generation" instead of "random generation"

* Initialize the library with a successful call to `psa_crypto_init()`.

> This example shows how to generate a random, ten-byte piece of data by calling `psa_generate_random()`:

Reviewer: Suggestion:
"This example shows how to generate a 10 bytes of random data `psa_generate_random()`:"

#### Handling hash operation contexts

> After a successful call to `psa_hash_setup()` initializes the operation structure, you can terminate the operation at any time by calling `psa_hash_abort()`. The call to `psa_hash_abort()` frees any resources associated with the operation, except for the operation structure itself.

Reviewer 2: Preexisting: please remove “initializes the operation structure”. `psa_hash_setup` does not actually initialize the operation structure. “Initialize” and “setup” are different steps, but `psa_hash_setup` did both in the past and we hadn't updated the wording here.

Reviewer: This needs updating still.

    /* Clean up */
    mbedtls_psa_crypto_free();
```

> To generate a random key, use `psa_generate_key()` instead of `psa_generate_random()`.

Reviewer: This is better placed at the beginning of this section, perhaps as a note or warning. Readers are likely to miss this if it is at the bottom and they followed along until they got what they thought they needed and stopped reading.

other data.

You must first initialize and set up a key derivation context,
provided with a key and, optionally, other data. Then, use the key derivation context to either read derived data to a buffer or send derived data directly to a key slot.

See the documentation for the particular algorithm (such as HKDF or the TLS1.2 PRF) for
information about which inputs to pass when, and when you can obtain which outputs.

**Prerequisites to working with the key derivation APIs:**
* Initialize the library with a successful call to `psa_crypto_init()`.
* Use a key with the appropriate attributes set:
    * Usage flags set for key derivation (`PSA_KEY_USAGE_DERIVE`)
    * Key type set to `PSA_KEY_TYPE_DERIVE`.
    * Algorithm set to a key derivation algorithm
      (`PSA_ALG_HKDF(PSA_ALG_SHA_256)`).

**To derive a new AES-CTR 128-bit encryption key into a given key slot using HKDF
with a given key, salt and information:**

1. Set up the key derivation context using the `psa_key_derivation_setup()`
function, specifying the derivation algorithm `PSA_ALG_HKDF(PSA_ALG_SHA_256)`.
1. Provide an optional salt with `psa_key_derivation_input_bytes()`.

> 1. Provide information with `psa_key_derivation_input_bytes()`.

Reviewer: We should put "info" in backticks. In this use case, it's a special name, not shorthand for "information".

> 1. Provide a secret with `psa_key_derivation_input_key()`, referencing a key that

Reviewer: We should put secret in backticks. In this use case, we refer to the variable name.

Mbed Crypto provides a key derivation API that lets you derive new keys from
existing ones. The key derivation API has functions to take inputs, including
other keys and data, and functions to generate outputs, such as new keys or
other data.

You must first initialize and set up a key derivation context,
provided with a key and, optionally, other data. Then, use the key derivation context to either read derived data to a buffer or send derived data directly to a key slot.

See the documentation for the particular algorithm (such as HKDF or the TLS1.2 PRF) for
information about which inputs to pass when, and when you can obtain which outputs.

**Prerequisites to working with the key derivation APIs:**
* Initialize the library with a successful call to `psa_crypto_init()`.
* Use a key with the appropriate attributes set:
    * Usage flags set for key derivation (`PSA_KEY_USAGE_DERIVE`)
    * Key type set to `PSA_KEY_TYPE_DERIVE`.
    * Algorithm set to a key derivation algorithm

>       (`PSA_ALG_HKDF(PSA_ALG_SHA_256)`).

Reviewer 2: Preexisting: the other two parenthetical remarks above are explanations, but this one is an example.
```suggestion
      (for example `PSA_ALG_HKDF(PSA_ALG_SHA_256)`).
```

**To derive a new AES-CTR 128-bit encryption key into a given key slot using HKDF

> with a given key, salt and information:**

Reviewer 2: The input is actually called “info”, not “information”. See https://tools.ietf.org/html/rfc5869#section-3.2

1. Set up the key derivation context using the `psa_key_derivation_setup()`
function, specifying the derivation algorithm `PSA_ALG_HKDF(PSA_ALG_SHA_256)`.
1. Provide an optional salt with `psa_key_derivation_input_bytes()`.
1. Provide `info` with `psa_key_derivation_input_bytes()`.
1. Provide `secret` with `psa_key_derivation_input_key()`, referencing a key that
   can be used for key derivation.
1. Set the key attributes desired for the new derived key. We'll set

>    the `PSA_KEY_USAGE_ENCRYPT` parameter and the `PSA_ALG_CTR` algorithm for this

Reviewer 2: ```suggestion
   the `PSA_KEY_USAGE_ENCRYPT` usage flag and the `PSA_ALG_CTR` algorithm for this
```

Mbed Crypto provides a key derivation API that lets you derive new keys from
existing ones. The key derivation API has functions to take inputs, including
other keys and data, and functions to generate outputs, such as new keys or
other data.

You must first initialize and set up a key derivation context,
provided with a key and, optionally, other data. Then, use the key derivation context to either read derived data to a buffer or send derived data directly to a key slot.

See the documentation for the particular algorithm (such as HKDF or the TLS1.2 PRF) for
information about which inputs to pass when, and when you can obtain which outputs.

**Prerequisites to working with the key derivation APIs:**
* Initialize the library with a successful call to `psa_crypto_init()`.
* Use a key with the appropriate attributes set:
    * Usage flags set for key derivation (`PSA_KEY_USAGE_DERIVE`)
    * Key type set to `PSA_KEY_TYPE_DERIVE`.
    * Algorithm set to a key derivation algorithm
      (for example `PSA_ALG_HKDF(PSA_ALG_SHA_256)`).

**To derive a new AES-CTR 128-bit encryption key into a given key slot using HKDF

> with a given key, salt and `info`:**

Reviewer: info should not be in backticks here. It's not short for information, nor referencing code. It's the name used in the encryption standard.

other keys and data, and functions to generate outputs, such as new keys or
other data.

You must first initialize and set up a key derivation context,
provided with a key and, optionally, other data. Then, use the key derivation context to either read derived data to a buffer or send derived data directly to a key slot.

See the documentation for the particular algorithm (such as HKDF or the TLS1.2 PRF) for
information about which inputs to pass when, and when you can obtain which outputs.

**Prerequisites to working with the key derivation APIs:**
* Initialize the library with a successful call to `psa_crypto_init()`.
* Use a key with the appropriate attributes set:
    * Usage flags set for key derivation (`PSA_KEY_USAGE_DERIVE`)
    * Key type set to `PSA_KEY_TYPE_DERIVE`.
    * Algorithm set to a key derivation algorithm
      (for example, `PSA_ALG_HKDF(PSA_ALG_SHA_256)`).

**To derive a new AES-CTR 128-bit encryption key into a given key slot using HKDF
with a given key, salt and info:**

1. Set up the key derivation context using the `psa_key_derivation_setup()`
function, specifying the derivation algorithm `PSA_ALG_HKDF(PSA_ALG_SHA_256)`.

> 1. Provide an optional salt with `psa_key_derivation_input_bytes()`.

Reviewer: For consistency, may want to put "salt" into backticks, or remove backticks from the "info" and "secret" references below.

Reviewer 2: Oh indeed there shouldn't be backticks around `info`. It isn't a name from the API, it's a word in English technical jargon.

1. Provide `info` with `psa_key_derivation_input_bytes()`.
1. Provide `secret` with `psa_key_derivation_input_key()`, referencing a key that
   can be used for key derivation.
1. Set the key attributes desired for the new derived key. We'll set
   the `PSA_KEY_USAGE_ENCRYPT` usage flag and the `PSA_ALG_CTR` algorithm for this
   example.
1. Derive the key by calling `psa_key_derivation_output_key()`.
1. Clean up the key derivation context.

At this point, the derived key slot holds a new 128-bit AES-CTR encryption key

> derived from the key, salt and `info` provided:

Reviewer: Remove backticks from "info" here. It's a term from the encryption standard, not an abbreviation for "information" or a reference to the code.

1. Set up the key derivation context using the `psa_key_derivation_setup()`
function, specifying the derivation algorithm `PSA_ALG_HKDF(PSA_ALG_SHA_256)`.

> 1. Provide `salt` (optional) with `psa_key_derivation_input_bytes()`.

Reviewer 2: No, “an optional salt” was correct.

Mbed Crypto provides a simple way to generate a key or key pair.

**Prerequisites to using key generation and export APIs:**
* Initialize the library with a successful call to `psa_crypto_init()`.

**To generate an ECDSA key:**
1. Set the desired key attributes for key generation by calling
   `psa_set_key_algorithm()` with the chosen ECDSA algorithm (such as
   `PSA_ALG_DETERMINISTIC_ECDSA(PSA_ALG_SHA_256)`). Do not set

>    `PSA_KEY_USAGE_EXPORT` because we only want to export the public key, not the key

Reviewer: Maybe replace "because" with a ";". This is too easy to read as "Don't set it because of some reason (you may set it for some other reason)" instead of "Don't set it. Here's why you shouldn't set it."

Mbed Crypto provides a simple way to generate a key or key pair.

**Prerequisites to using key generation and export APIs:**
* Initialize the library with a successful call to `psa_crypto_init()`.

**To generate an ECDSA key:**
1. Set the desired key attributes for key generation by calling
   `psa_set_key_algorithm()` with the chosen ECDSA algorithm (such as
   `PSA_ALG_DETERMINISTIC_ECDSA(PSA_ALG_SHA_256)`). Do not set

>    `PSA_KEY_USAGE_EXPORT`; we only want to export the public key, not the key

Reviewer 2: “Do not” is inconsistent with “we want”. Also I think a semicolon is not clear here. What follows is an explanation, so a colon would work, or a word like “since”.

    mbedtls_psa_crypto_free();
```

> ### More about the Mbed Crypto

Reviewer: Replace "More about the Mbed Crypto" with "More about the PSA Crypto API" or "More about PSA Crypto".

    mbedtls_psa_crypto_free();
```

> ### More about the Mbed Crypto API

Reviewer 2: The Mbed Crypto library doesn't just implement the PSA Crypto API, it also implements the crypto part of the Mbed TLS API. The title of this section doesn't match the content anymore.

> For more information about PSA Crypto, download the *PSA Cryptography API* PDF under [PSA APIs](https://developer.arm.com/architectures/security-architectures/platform-security-architecture#implement).

Reviewer 2: This URL is the official PDF which is not up-to-date. It would be better to link to the copy that's more useful to developers, which has moved to https://github.com/ARMmbed/mbed-crypto/raw/psa-crypto-api/docs/PSA_Cryptography_API_Specification.pdf and also has an HTML version at https://armmbed.github.io/mbed-crypto/html/index.html

    mbedtls_psa_crypto_free();
```

> ### More about the Mbed Crypto API

Reviewer: Replace "Mbed" with "PSA" here
